pkg/cloudinary/collector.go

```diff
@@ -0,0 +1,227 @@
+				ConstLabels: nil,
+			},
+		),
+		storage_limit_bytes: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   "cloudinary",
+				Name:        "storage_limit_bytes",
+				Help:        "Limit of storage used allowed in the last 30 days",
+				ConstLabels: nil,
+			},
+		),
+		storage_usage_ratio: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   "cloudinary",
+				Name:        "storage_usage_ratio",
+				Help:        "Ratio of storage bytes used over corresponding limit",
+				ConstLabels: nil,
+			},
+		),
+		requests_amount: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   "cloudinary",
+				Name:        "requests_amount",
+				Help:        "Number of requests done to Cloudinary",
+				ConstLabels: nil,
+			},
+		),
+		resources_amount: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   "cloudinary",
+				Name:        "resources_amount",
+				Help:        "Number of resources in Cloudinary",
+				ConstLabels: nil,
+			},
+		),
+		derived_resources_amount: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   "cloudinary",
+				Name:        "derived_resources_amount",
+				Help:        "Number of derived resources in Cloudinary",
+				ConstLabels: nil,
+			},
+		),
+	}
+}
+
+// Describe is a requirement for the Collector interface of Prometheus
+// that returns each exported metric's description to the Prometheus
+// middleware
+func (c *Collector) Describe(ch chan<- *prometheus.Desc) {
+	c.transformations_usage_amount.Describe(ch)
+	c.transformations_limit_amount.Describe(ch)
+	c.transformations_usage_ratio.Describe(ch)
+	c.objects_usage_amount.Describe(ch)
+	c.objects_limit_amount.Describe(ch)
+	c.objects_usage_ratio.Describe(ch)
+	c.bandwidth_usage_bytes.Describe(ch)
+	c.bandwidth_limit_bytes.Describe(ch)
+	c.bandwidth_usage_ratio.Describe(ch)
+	c.storage_usage_bytes.Describe(ch)
+	c.storage_limit_bytes.Describe(ch)
+	c.storage_usage_ratio.Describe(ch)
+	c.requests_amount.Describe(ch)
+	c.resources_amount.Describe(ch)
+	c.derived_resources_amount.Describe(ch)
+}
+
+// Collect is a requirement for the Collector interface of Prometheus
+// that runs the queries to set the metrics values to be exported
+func (c *Collector) Collect(ch chan<- prometheus.Metric) {
+	c.runCollection()
+
+	c.transformations_usage_amount.Collect(ch)
+	c.transformations_limit_amount.Collect(ch)
+	c.transformations_usage_ratio.Collect(ch)
+	c.objects_usage_amount.Collect(ch)
+	c.objects_limit_amount.Collect(ch)
+	c.objects_usage_ratio.Collect(ch)
+	c.bandwidth_usage_bytes.Collect(ch)
+	c.bandwidth_limit_bytes.Collect(ch)
+	c.bandwidth_usage_ratio.Collect(ch)
+	c.storage_usage_bytes.Collect(ch)
+	c.storage_limit_bytes.Collect(ch)
+	c.storage_usage_ratio.Collect(ch)
+	c.requests_amount.Collect(ch)
+	c.resources_amount.Collect(ch)
+	c.derived_resources_amount.Collect(ch)
+}
+
+func (c *Collector) runCollection() {
+	cloudName, key, secret := CloudinaryCredentials.get()
```

Comment:
Why do you need the `.get()`, can't you just unpack it manually 4 lines further down?

Reply:
Because the variables are private to the struct they can't be accessed manually

Reply:
Go's spec allows to directly unpack on variadic argumentlists, but not mixing static and variadic, so, in that case, working code would make get return the URL as well, which is completely out of the `CloudinaryCredentials` responsability.
Anyway, current credentials management rightness is not clear yet, and this might change in the future.